File: sdk-api/src/main/java/dev/restate/sdk/Service.java
// Copyright (c) 2023 - Restate Software, Inc., Restate GmbH
//
// This file is part of the Restate Java SDK,
// which is released under the MIT license.
//
// You can find a copy of the license in file LICENSE in the root
// directory of this repository or package, or at
// https://github.com/restatedev/sdk-java/blob/main/LICENSE
package dev.restate.sdk;

import com.google.protobuf.ByteString;
import dev.restate.sdk.common.*;
import dev.restate.sdk.common.ServiceType;
import dev.restate.sdk.common.syscalls.*;
import io.opentelemetry.context.Scope;
import java.util.*;
import java.util.concurrent.Executor;
import java.util.concurrent.Executors;
import java.util.function.*;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

public final class Service implements BindableService<Service.Options> {
  private final ServiceDefinition<Options> serviceDefinition;
  private final Service.Options options;

  private Service(
      String fqsn,
      boolean isKeyed,
      HashMap<String, HandlerDefinition<?, ?, Options>> handlers,
      Options options) {
    this.serviceDefinition =
        ServiceDefinition.of(
            fqsn,
            isKeyed ? ServiceType.VIRTUAL_OBJECT : ServiceType.SERVICE,
            new ArrayList<>(handlers.values()));
    this.options = options;
  }

  @Override
  public Options options() {
    return this.options;
  }

  @Override
  public List<ServiceDefinition<Options>> definitions() {
    return List.of(this.serviceDefinition);
  }

  public static ServiceBuilder service(String name) {
    return new ServiceBuilder(name);
  }

  public static VirtualObjectBuilder virtualObject(String name) {
    return new VirtualObjectBuilder(name);
  }

  public static class AbstractServiceBuilder {

    protected final String name;
    protected final HashMap<String, HandlerDefinition<?, ?, Options>> handlers;

    public AbstractServiceBuilder(String name) {
      this.name = name;
      this.handlers = new HashMap<>();
    }
  }

  public static class VirtualObjectBuilder extends AbstractServiceBuilder {

    VirtualObjectBuilder(String name) {
      super(name);
    }

    public <REQ, RES> VirtualObjectBuilder withShared(
        String name,
        Serde<REQ> requestSerde,
        Serde<RES> responseSerde,
        BiFunction<SharedObjectContext, REQ, RES> runner) {
      this.handlers.put(
          name,
          HandlerDefinition.of(
              HandlerSpecification.of(name, HandlerType.SHARED, requestSerde, responseSerde),
              new Handler<>(runner)));
      return this;
    }

    public <REQ, RES> VirtualObjectBuilder withExclusive(
        String name,
        Serde<REQ> requestSerde,
        Serde<RES> responseSerde,
        BiFunction<ObjectContext, REQ, RES> runner) {
      this.handlers.put(
          name,
          HandlerDefinition.of(
              HandlerSpecification.of(name, HandlerType.EXCLUSIVE, requestSerde, responseSerde),
              new Handler<>(runner)));
      return this;
    }

    public Service build(Service.Options options) {
      return new Service(this.name, true, this.handlers, options);
    }
  }

  public static class ServiceBuilder extends AbstractServiceBuilder {

    ServiceBuilder(String name) {
      super(name);
    }

    public <REQ, RES> ServiceBuilder with(
        String name,
        Serde<REQ> requestSerde,
        Serde<RES> responseSerde,
        BiFunction<Context, REQ, RES> runner) {
      this.handlers.put(
          name,
          HandlerDefinition.of(
              HandlerSpecification.of(name, HandlerType.SHARED, requestSerde, responseSerde),
              new Handler<>(runner)));
      return this;
    }

    public Service build(Service.Options options) {
      return new Service(this.name, false, this.handlers, options);
    }
  }

  public static class Handler<REQ, RES> implements InvocationHandler<REQ, RES, Service.Options> {
    private final BiFunction<Context, REQ, RES> runner;

    private static final Logger LOG = LogManager.getLogger(Handler.class);

    Handler(BiFunction<? extends Context, REQ, RES> runner) {
      //noinspection unchecked
      this.runner = (BiFunction<Context, REQ, RES>) runner;
    }

    public BiFunction<Context, REQ, RES> getRunner() {
      return runner;
    }

    @Override
    public void handle(
        HandlerSpecification<REQ, RES> handlerSpecification,
        Syscalls syscalls,
        Service.Options options,
        SyscallCallback<ByteString> callback) {
      // Wrap the executor for setting/unsetting the thread local
      Executor wrapped =
          runnable ->
              options.executor.execute(
                  () -> {
                    SYSCALLS_THREAD_LOCAL.set(syscalls);
                    try (Scope ignored = syscalls.request().otelContext().makeCurrent()) {
                      runnable.run();
                    } finally {
                      SYSCALLS_THREAD_LOCAL.remove();
                    }
                  });
      wrapped.execute(
          () -> {
            // Any context switching, if necessary, will be done by ResolvedEndpointHandler
            Context ctx = new ContextImpl(syscalls);

            // Parse input
            REQ req;
            try {
              req =
                  handlerSpecification
                      .getRequestSerde()
                      .deserialize(syscalls.request().bodyBuffer());
            } catch (Error e) {
              throw e;
            } catch (Throwable e) {
              LOG.warn("Cannot deserialize input", e);
              callback.onCancel(
                  new TerminalException(
                      TerminalException.BAD_REQUEST_CODE,
                      "Cannot deserialize input: " + e.getMessage()));
              return;
            }

            // Execute user code
            RES res;
            try {
              res = this.runner.apply(ctx, req);
            } catch (Error e) {
              throw e;
            } catch (Throwable e) {
              callback.onCancel(e);
              return;
            }

            // Serialize output
            ByteString serializedResult;
            try {
              serializedResult = handlerSpecification.getResponseSerde().serializeToByteString(res);
            } catch (Error e) {
              throw e;
            } catch (Throwable e) {
              LOG.warn("Cannot serialize output", e);
              callback.onCancel(
                  new TerminalException(
                      TerminalException.INTERNAL_SERVER_ERROR_CODE,
                      "Cannot serialize output: " + e.getMessage()));
              return;
            }

            // Complete callback
            callback.onSuccess(serializedResult);
          });
    }

    public static <CTX extends Context, REQ, RES> Handler<REQ, RES> of(BiFunction<CTX, REQ, RES> runner) {
      return new Handler<>(runner);
    }

    @SuppressWarnings("unchecked")
    public static <CTX extends Context, RES> Handler<Void, RES> of(Function<CTX, RES> runner) {
      return new Handler<>((context, o) -> runner.apply((CTX) context));
    }

    @SuppressWarnings("unchecked")
    public static <CTX extends Context, REQ> Handler<REQ, Void> of(BiConsumer<CTX, REQ> runner) {
      return new Handler<>((context, o) -> {
        runner.accept((CTX) context, o);
        return null;
      });
    }

    @SuppressWarnings("unchecked")
    public static <CTX extends Context> Handler<Void, Void> of(Consumer<CTX> runner) {
      return new Handler<>((ctx, o) -> {
        runner.accept((CTX) ctx);
        return null;
      });
    }
  }

  public static class Options {
    public static final Options DEFAULT = new Options(Executors.newCachedThreadPool());

    private final Executor executor;

    /**
     * You can run on virtual threads by using the executor {@code
     * Executors.newVirtualThreadPerTaskExecutor()}.
     */
    public Options(Executor executor) {
      this.executor = executor;
    }
  }
}

File: sdk-api-kotlin/src/main/kotlin/dev/restate/sdk/kotlin/Service.kt
// Copyright (c) 2023 - Restate Software, Inc., Restate GmbH
//
// This file is part of the Restate Java SDK,
// which is released under the MIT license.
//
// You can find a copy of the license in file LICENSE in the root
// directory of this repository or package, or at
// https://github.com/restatedev/sdk-java/blob/main/LICENSE
package dev.restate.sdk.kotlin

import com.google.protobuf.ByteString
import dev.restate.sdk.common.*
import dev.restate.sdk.common.syscalls.*
import io.opentelemetry.extension.kotlin.asContextElement
import kotlin.coroutines.CoroutineContext
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.asContextElement
import kotlinx.coroutines.launch
import org.apache.logging.log4j.LogManager

class Service
private constructor(
    fqsn: String,
    isKeyed: Boolean,
    handlers: Map<String, HandlerDefinition<*, *, Options>>,
    private val options: Options
) : BindableService<Service.Options> {
  private val serviceDefinition =
      ServiceDefinition(
          fqsn,
          if (isKeyed) ServiceType.VIRTUAL_OBJECT else ServiceType.SERVICE,
          handlers.values.toList())

  override fun options(): Options {
    return this.options
  }

  override fun definitions() = listOf(this.serviceDefinition)

  companion object {
    fun service(
        name: String,
        options: Options = Options.DEFAULT,
        init: ServiceBuilder.() -> Unit
    ): Service {
      val builder = ServiceBuilder(name)
      builder.init()
      return builder.build(options)
    }

    fun virtualObject(
        name: String,
        options: Options = Options.DEFAULT,
        init: VirtualObjectBuilder.() -> Unit
    ): Service {
      val builder = VirtualObjectBuilder(name)
      builder.init()
      return builder.build(options)
    }
  }

  class VirtualObjectBuilder internal constructor(private val name: String) {
    private val handlers: MutableMap<String, HandlerDefinition<*, *, Options>> = mutableMapOf()

    fun <REQ, RES> sharedHandler(
        name: String,
        requestSerde: Serde<REQ>,
        responseSerde: Serde<RES>,
        runner: suspend (SharedObjectContext, REQ) -> RES
    ): VirtualObjectBuilder {
      handlers[name] =
          HandlerDefinition(
              HandlerSpecification.of(name, HandlerType.SHARED, requestSerde, responseSerde),
              Handler(runner))
      return this
    }

    inline fun <reified REQ, reified RES> sharedHandler(
        name: String,
        noinline runner: suspend (SharedObjectContext, REQ) -> RES
    ) = this.sharedHandler(name, KtSerdes.json(), KtSerdes.json(), runner)

    fun <REQ, RES> exclusiveHandler(
        name: String,
        requestSerde: Serde<REQ>,
        responseSerde: Serde<RES>,
        runner: suspend (ObjectContext, REQ) -> RES
    ): VirtualObjectBuilder {
      handlers[name] =
          HandlerDefinition(
              HandlerSpecification.of(name, HandlerType.EXCLUSIVE, requestSerde, responseSerde),
              Handler(runner))
      return this
    }

    inline fun <reified REQ, reified RES> exclusiveHandler(
        name: String,
        noinline runner: suspend (ObjectContext, REQ) -> RES
    ) = this.exclusiveHandler(name, KtSerdes.json(), KtSerdes.json(), runner)

    fun build(options: Options) = Service(this.name, true, this.handlers, options)
  }

  class ServiceBuilder internal constructor(private val name: String) {
    private val handlers: MutableMap<String, HandlerDefinition<*, *, Options>> = mutableMapOf()

    fun <REQ, RES> handler(
        name: String,
        requestSerde: Serde<REQ>,
        responseSerde: Serde<RES>,
        runner: suspend (SharedObjectContext, REQ) -> RES
    ): ServiceBuilder {
      handlers[name] =
          HandlerDefinition(
              HandlerSpecification.of(name, HandlerType.SHARED, requestSerde, responseSerde),
              Handler(runner))
      return this
    }

    inline fun <reified REQ, reified RES> handler(
        name: String,
        noinline runner: suspend (Context, REQ) -> RES
    ) = this.handler(name, KtSerdes.json(), KtSerdes.json(), runner)

    fun build(options: Options) = Service(this.name, false, this.handlers, options)
  }

  class Handler<REQ, RES, CTX : Context>(
      private val runner: suspend (CTX, REQ) -> RES,
  ) : InvocationHandler<REQ, RES, Options> {

    companion object {
      private val LOG = LogManager.getLogger()
    }

    override fun handle(
        handlerSpecification: HandlerSpecification<REQ, RES>,
        syscalls: Syscalls,
        options: Options,
        callback: SyscallCallback<ByteString>
    ) {
      val ctx: Context = ContextImpl(syscalls)

      val scope =
          CoroutineScope(
              options.coroutineContext +
                  InvocationHandler.SYSCALLS_THREAD_LOCAL.asContextElement(syscalls) +
                  syscalls.request().otelContext()!!.asContextElement())
      scope.launch {
        val serializedResult: ByteString

        try {
          // Parse input
          val req: REQ
          try {
            req = handlerSpecification.requestSerde.deserialize(syscalls.request().bodyBuffer())
          } catch (e: Error) {
            throw e
          } catch (e: Throwable) {
            LOG.warn("Error when deserializing input", e)
            throw TerminalException(
                TerminalException.BAD_REQUEST_CODE, "Cannot deserialize input: " + e.message)
          }

          // Execute user code
          @Suppress("UNCHECKED_CAST") val res: RES = runner(ctx as CTX, req)

          // Serialize output
          try {
            serializedResult = handlerSpecification.responseSerde.serializeToByteString(res)
          } catch (e: Error) {
            throw e
          } catch (e: Throwable) {
            LOG.warn("Error when serializing input", e)
            throw TerminalException(
                TerminalException.INTERNAL_SERVER_ERROR_CODE, "Cannot serialize output: $e")
          }
        } catch (e: Throwable) {
          callback.onCancel(e)
          return@launch
        }

        // Complete callback
        callback.onSuccess(serializedResult)
      }
    }
  }

  class Options(val coroutineContext: CoroutineContext) {
    companion object {
      val DEFAULT: Options = Options(Dispatchers.Default)
    }
  }
}
